README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cronologia

<!-- badges: start -->

[![R-CMD-check](https://github.com/feddelegrand7/cronologia/workflows/R-CMD-check/badge.svg)](https://github.com/feddelegrand7/cronologia/actions)
[![Codecov test
coverage](https://codecov.io/gh/feddelegrand7/cronologia/branch/master/graph/badge.svg)](https://codecov.io/gh/feddelegrand7/cronologia?branch=master)
[![License: AGPL
v3](https://img.shields.io/badge/License-AGPL%20v3-blue.svg)](https://www.gnu.org/licenses/agpl-3.0)
[![CRAN
status](https://www.r-pkg.org/badges/version/cronologia)](https://CRAN.R-project.org/package=cronologia)
[![metacran
downloads](https://cranlogs.r-pkg.org/badges/cronologia)](https://cran.r-project.org/package=cronologia)
[![metacran
downloads](https://cranlogs.r-pkg.org/badges/grand-total/cronologia)](https://cran.r-project.org/package=cronologia)
[![R
badge](https://img.shields.io/badge/Build%20with-♥%20and%20R-violet)](https://github.com/feddelegrand7/cronologia)

<!-- badges: end -->

The goal of `cronologia` is to create an interactive timeline widget in
RMarkdown documents and Shiny applications.

## Installation

You can install the stable version from CRAN with

``` r
install.packages("cronologia")
```

# Introduction

The `cronologia` package has three functions:

-   `create_tml()` : used to create simple text-based timelines.
-   `create_tml_img()`: used to create timelines that include images.
-   `create_tml_2()`: works the same way as `create_tml()` except that
    it adds and additional description component.

# Examples

## `create_tml()`

In order to showcase the package’s features, let’s create a simple data
frame:

``` r
batman_data <- data.frame(


  date_release = c("May 31, 2005",
                   "July 14, 2008",
                   "July 16, 2012 "),

  title = c("Batman Begins",
                  "The Dark Knight",
                  "The Dark Knight Rises")
)

batman_data
#>     date_release                 title
#> 1   May 31, 2005         Batman Begins
#> 2  July 14, 2008       The Dark Knight
#> 3 July 16, 2012  The Dark Knight Rises
```

Now, using `create_tml()`, we can create easily a timeline as follows:

``` r
library(cronologia)


create_tml(df = batman_data, # the data frame
           smr = "title", # the column that will be used in the summary 
           dsc = "date_release" # the column that will be used in the description
           )
```

![](man/figures/example1.gif)

You can easily customize the appearance of the time line using the
parameters provided:

``` r
create_tml(df = batman_data,
           smr = "title", # summary
           dsc = "date_release", # description
           smr_col = "blue", # summary text color
           smr_bgcol = "orange", # summary background color
           dsc_col = "white", # description text color
           dsc_bgcol = "black", # description background color
           dsc_size = "30px" # description size
           )
```

![](man/figures/example2.gif)

### ❗❗❗

If you want to make all the summary components open by default, you can
set the `open` parameter to `TRUE`. The parameter is available in all
the functions.

## `create_tml_img()`

If you want to include images within your timeline, you can use the
`create_tml_img()` function. To illustrate this function, we’ll use the
[radous](https://github.com/feddelegrand7/radous) package that fetch the
[randomuser.me](https://randomuser.me/) API and returns a data frame
that contains many information (including images’ URLs).

> Disclaimer: All the generated images are extracted from the authorized
> section of UI Faces.

``` r
library(radous)

df <- get_data(n = 4, seed = "123")

df[c('name_last', 
     'location_street_name',
     'picture_large',
     'name_last')]
#> # A tibble: 4 x 4
#>   name_last location_street_na~ picture_large                          name_last
#>   <chr>     <chr>               <chr>                                  <chr>    
#> 1 Campos    Rua Três            https://randomuser.me/api/portraits/m~ Campos   
#> 2 Jackson   Armagh Street       https://randomuser.me/api/portraits/w~ Jackson  
#> 3 Ruona     Hämeentie           https://randomuser.me/api/portraits/w~ Ruona    
#> 4 Steward   Henry Street        https://randomuser.me/api/portraits/w~ Steward
```

Now we will proceed as previously except that we need to provide two
additional arguments:

-   `imgsrc`: the column that indicates the source of the images.
-   `imgalt`: the column indicating the `alt` attribute of the images.
    For accessibility reasons I decided to make this argument mandatory.
    Use a column that contains `""` if the images do not need the `alt`
    attribute.

``` r
df <- radous::get_data(4, seed = "123")

create_tml_img(df, 
               smr = "name_last", 
               dsc = "location_street_name", 
               imgsrc = "picture_large", 
               imgalt = "name_last", 
               imgwidth = "150px", 
               imgheight = "150px", 
               dsc_size = "20px")
```

![](man/figures/example3.gif)

## `create_tml_2()`

Following the idea of
[Tobias](https://twitter.com/toeb18/status/1355104693299634181?s=20) for
creating an interactive CV I thought that two (2) description components
would be more appropriate. The function is similar to `create_tml()`
except that it adds another description paragraph to the Timeline.

Let’s go through an example:

``` r
cv <- data.frame(
  
  
  jobs = c("Game tester at Nintendo", "Food tester at Ferrero", "Movies tester at Netflix"), 
  
  period = c("2020-2022", "2022-2024", "2026-2030"),
  
  todos = c("Playing Zelda all day", "Eating Bueno all day", "Watching the Office all day")
  
)


cv
#>                       jobs    period                       todos
#> 1  Game tester at Nintendo 2020-2022       Playing Zelda all day
#> 2   Food tester at Ferrero 2022-2024        Eating Bueno all day
#> 3 Movies tester at Netflix 2026-2030 Watching the Office all day
```

``` r
create_tml_2(cv, 
             smr = "jobs", 
             dsc = "period", 
             dsc2 = "todos", 
             dsc2_col = "white",
             dsc2_bgcol = "peru") # yes, peru is also a color
```

![](man/figures/example4.gif)

# TODOS

-   [x] Writing unit tests
-   [ ] Creating a hex sticker
-   [x] Sharing with RWeekly
-   [ ] Talk about it at a virtual event

# Code of Conduct

Please note that the `cronologia` project is released with a
[Contributor Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# cronologia 0.2.0

* I've added CSS responsiveness for smaller devices. 

# cronologia 0.1.0

* First version to CRAN. 


tests/testthat.R
library(testthat)
library(cronologia)

test_check("cronologia")


tests/testthat/test-cronologia.R

test_that("expect create_tml() to return an error when
          a column does not belong to a dataframe
          ", {


  expect_error(

    create_tml(mtcars,
               smr = "mpg",
               dsc = "tacos")

  )

  expect_error(

    create_tml(mtcars,
               smr = "tacos",
               dsc = "mpg")

  )

          })



test_that("expect create_tml_img() to return an error when
          a column does not belong to a dataframe", {


            expect_error(

              create_tml_img(mtcars,
                             smr = "mpg",
                             dsc = "wt",
                             imgsrc = "tacos",
                             imgalt = "disp")

            )

            expect_error(

              create_tml_img(mtcars,
                             smr = "tacos",
                             dsc = "mpg",
                             imgalt = "wt",
                             imgsrc = "qsec"
                            )

            )

            expect_error(

              create_tml_img(mtcars,
                             smr = "cyl",
                             dsc = "mpg",
                             imgalt = "tacos",
                             imgsrc = "drat"
              )

            )

          })


test_that("create_tml() returns an object of class shiny.tag.list", {


  expect_s3_class(

    object = create_tml(mtcars, "mpg", "wt"),

    class = "shiny.tag.list"


  )


})



test_that("create_tml_img() returns an object of class shiny.tag.list", {


  expect_s3_class(

    object = create_tml_img(mtcars,
                            "mpg",
                            "wt",
                            "drat",
                            "disp"),

    class = "shiny.tag.list"


  )


})


test_that("create_tml() returns an object of class list", {


  expect_s3_class(

    object = create_tml(mtcars, "mpg", "wt"),

    class = "list"


  )


})


test_that("create_tml_img() returns an object of class list", {


  expect_s3_class(

    object = create_tml_img(mtcars,
                            "mpg",
                            "wt",
                            "drat",
                            "disp"),

    class = "list"


  )


})


test_that("create_tml() returns an object of type list", {


  expect_type(

    object = create_tml(mtcars, "mpg", "wt"),

    type = "list"


  )


})



test_that("create_tml_img() returns an object of type list", {


  expect_type(

    object = create_tml_img(mtcars,
                            "mpg",
                            "wt",
                            "drat",
                            "disp"),

    type = "list"


  )


})


test_that("create_tml() returns an error when arguments missing", {




  expect_error(

    create_tml(mtcars, smr = "mpg")

  )

  expect_error(

    create_tml(smr = "mpg", dsc = "wt")

  )




})


test_that("create_tml_img() returns an error when arguments missing", {




  expect_error(

    create_tml_img(mtcars,
                   smr = "mpg",
                   dsc = "wt",
                   imgalt = "disp")

  )

  expect_error(

    create_tml_img(mtcars,
                   smr = "mpg",
                   dsc = "wt",
                   imgsrc = "disp")
  )




})


R/cronologia.R

#' An internal function used to manage the CSS file dependency
#' @noRd
cronologia_dependency <- function() {

  htmltools::htmlDependency(

    name = "cronologia",

    version = "0.1.0",

    package = "cronologia",

    src = c(file = "cronologia"),

    stylesheet = "cronologia.min.css",

    all_files = FALSE

  )

}

#' An internal function used to generate HTML tags for the timeline function
#' @noRd
timeline <- function(df,
                     smr,
                     dsc,
                     smr_col,
                     smr_bgcol,
                     dsc_col,
                     dsc_bgcol,
                     dsc_size,
                     open) {




    func <- function(i, j) {


    det <-  htmltools::tags$details(

        htmltools::tags$summary(
          i,
          style = glue::glue(

            "color: {smr_col}; background: {smr_bgcol};"

          )
        ),

        htmltools::tags$p(
          j,

          style = glue::glue(

            "color: {dsc_col}; background: {dsc_bgcol}; font-size: {dsc_size};"

          )

        )
      )



    details <- if (open) {


      htmltools::tagAppendAttributes(det, "open" = NA)


    } else {

      det

    }

        htmltools::tagList(

          htmltools::div(
            class = "chronologia-detail-wrapper",

            details

          )
        )
    }

    r <- mapply(func, df[[smr]], df[[dsc]], SIMPLIFY = FALSE)

    return(

      htmltools::tagList(
        htmltools::div(class = "chronologia-wrapper", r)
      )
    )


}


#' Create a Vertical HTML Timeline from a Data Frame
#'
#' @param df a data frame containing the necessary information.
#' @param smr the column name that will be used as the summary component.
#' @param dsc the column name that will be used as the description component.
#' @param smr_col text color for the summary component. Defaults to 'white'.
#' @param smr_bgcol background-color for the summary component. Defaults to "#002240".
#' @param dsc_col text color for the description component. Defaults to "#1D1D1D".
#' @param dsc_bgcol background-color for the description component. Defaults to "#FFFFFF".
#' @param dsc_size font size of the description component. Defaults to "14px".
#' @param open logical, whether to open by default the summary components. Default to FALSE
#'
#'
#' @return A Vertical HTML Time Line Widget
#' @export
#'
#' @examples
#'
#' if(interactive()) {
#'
#'df <- data.frame(
#'
#'
#'  date_release = c("May 31, 2005",
#'                   "July 14, 2008",
#'                   "July 16, 2012 "),
#'
#'  description = c("Batman Begins",
#'                  "The Dark Knight",
#'                  "The Dark Knight Rises")
#')
#'
#'ui <- fluidPage(
#'
#'  br(),
#'
#'  h1("Batman Trigoly Timeline", style = "text-align:center"),
#'
#'  br(),
#'
#'
#'  create_tml(df,
#'            "date_release",
#'            "description"),
#'
#'
#'
#')
#'
#'server <- function(input, output) {}
#'
#'shinyApp(ui = ui, server = server)
#'
#'}


create_tml <- function(df,
                       smr,
                       dsc,
                       smr_col = "white",
                       smr_bgcol =  "#002240",
                       dsc_col = "#1D1D1D",
                       dsc_bgcol = "#FFFFFF",
                       dsc_size = "14px",
                       open = FALSE
                      ) {

  if (any(!is.character(smr),
          !is.character(dsc),
          !is.character(smr_col),
          !is.character(smr_bgcol),
          !is.character(dsc_col),
          !is.character(dsc_bgcol),
          !is.character(dsc_size))) {


    stop("All parameters except 'df' must be provided as a character string ! you're not in Tidyverse !!!")

  }


  if(any(!c(smr, dsc) %in% names(df))) {

    stop(glue::glue(

      "Please check the columns' names.

      Note that they need to be provided as a character string in this form:

      (df = mtcars, smr = 'mpg', dsc = 'wt')

      "

    ))

  }


  htmltools::tagList(

    cronologia_dependency(),

    timeline(df,
             smr,
             dsc,
             smr_col,
             smr_bgcol,
             dsc_col,
             dsc_bgcol,
             dsc_size,
             open)

  )

}



# Create Time Line with Images --------------------------------------------


#' An internal function used to create HTML tags for the timeline_img function
#' @noRd
timeline_img <- function(df,
                     smr,
                     dsc,
                     imgsrc,
                     imgalt,
                     smr_col,
                     smr_bgcol,
                     dsc_col,
                     dsc_bgcol,
                     dsc_size,
                     imgheight,
                     imgwidth,
                     open) {






  func <- function(i, j, n, m) {


    det <- htmltools::tags$details(

      htmltools::tags$summary(
        i,
        style = glue::glue(

          "color: {smr_col}; background: {smr_bgcol};"

        )
      ),

      htmltools::tags$img(
        src = n,
        alt = m,
        height = imgheight,
        width = imgwidth
      ),

      htmltools::tags$br(),
      htmltools::tags$br(),


      htmltools::tags$p(
        j,

        style = glue::glue(

          "color: {dsc_col}; background: {dsc_bgcol}; font-size: {dsc_size};"

        )

      )
    )


   details <-  if (open) {


     htmltools::tagAppendAttributes(det, "open" = NA)


    } else {

        det

    }


    htmltools::tagList(

      htmltools::div(
        class = "chronologia-detail-wrapper",
        details

      )
    )
  }

  r <- mapply(func,
              df[[smr]],
              df[[dsc]],
              df[[imgsrc]],
              df[[imgalt]],
              SIMPLIFY = FALSE)

  return(

    htmltools::tagList(
      htmltools::div(class = "chronologia-wrapper", r)
    )
  )


}




#' Create Vertical HTML Timeline Widget with Images from a Data Frame
#'
#' @param df a data frame containing the necessary information.
#' @param smr the column name that will be used as the summary component.
#' @param dsc the column name that will be used as the description component.
#' @param imgsrc the column name indicating the 'href' source of the images.
#' @param imgalt the column name that will be used as 'alt' text for the images.
#' @param smr_col text color for the summary component. Defaults to 'white'.
#' @param smr_bgcol background-color for the summary component. Defaults to "#002240".
#' @param dsc_col text color for the description component. Defaults to "#1D1D1D".
#' @param dsc_bgcol background-color for the description component. Defaults to "#FFFFFF".
#' @param dsc_size font size of the description component. Defaults to "14px".
#' @param imgheight the height of the image. Defaults to "150px".
#' @param imgwidth the width of the image. Defaults to "150px".
#' @param open logical, whether to open by default the summary components. Default to FALSE

#' @return A Vertical HTML Time Line Widget With Images
#' @export


create_tml_img <- function(df,
                       smr,
                       dsc,
                       imgsrc,
                       imgalt,
                       smr_col = "white",
                       smr_bgcol =  "#002240",
                       dsc_col = "#1D1D1D",
                       dsc_bgcol = "#FFFFFF",
                       dsc_size = "14px",
                       imgheight = "150px",
                       imgwidth = "150px",
                       open = FALSE) {


  if (any(!is.character(smr),
          !is.character(dsc),
          !is.character(imgsrc),
          !is.character(imgalt),
          !is.character(smr_col),
          !is.character(smr_bgcol),
          !is.character(dsc_col),
          !is.character(dsc_bgcol),
          !is.character(dsc_size),
          !is.character(imgheight),
          !is.character(imgwidth))) {


    stop("All parameters except 'df' must be provided as a character string ! you're not in Tidyverse !!!")

  }


  if(any(!c(smr, dsc, imgsrc, imgalt) %in% names(df))) {

    stop(glue::glue(

      "Please check the columns' names.

      Note that they need to be provided as a character string in this form:

      (df = mtcars, smr = 'mpg', dsc = 'wt', imgsrc = 'disp', imgalt = 'drat')

      "

    ))

  }


  htmltools::tagList(

    cronologia_dependency(),

    timeline_img(df,
             smr,
             dsc,
             imgsrc,
             imgalt,
             smr_col,
             smr_bgcol,
             dsc_col,
             dsc_bgcol,
             dsc_size,
             imgheight,
             imgwidth,
             open)

  )

}



# cronologia The Dark Knight ------------------------------------------------------------




#' An internal function used to generate HTML tags for the timeline2 function
#' @noRd

timeline2 <- function(df,
                     smr,
                     dsc,
                     dsc2,
                     smr_col,
                     smr_bgcol,
                     dsc_col,
                     dsc_bgcol,
                     dsc_size,
                     dsc2_col,
                     dsc2_bgcol,
                     dsc2_size,
                     open) {

  func <- function(i, j, k) {

    det <-   htmltools::tags$details(

      htmltools::tags$summary(
        i,
        style = glue::glue(

          "color: {smr_col}; background: {smr_bgcol};"

        )
      ),

      htmltools::tags$p(
        j,

        style = glue::glue(

          "color: {dsc_col}; background: {dsc_bgcol}; font-size: {dsc_size};"

        )

      ),


      htmltools::tags$p(

        k,

        style = glue::glue(

          "color: {dsc2_col}; background: {dsc2_bgcol}; font-size: {dsc2_size};"

        )


      )
    )



  details <-   if (open) {

    htmltools::tagAppendAttributes(det, "open" = NA)


  } else {

    det

    }


    htmltools::tagList(

      htmltools::div(
        class = "chronologia-detail-wrapper",

        details


      )
    )
  }

  r <- mapply(func, df[[smr]], df[[dsc]], df[[dsc2]], SIMPLIFY = FALSE)

  return(

    htmltools::tagList(
      htmltools::div(class = "chronologia-wrapper", r)
    )
  )


}


#' Create a Vertical HTML Timeline with 2 Description Levels from a Data Frame
#'
#' @param df a data frame containing the necessary information.
#' @param smr the column name that will be used as the summary component.
#' @param dsc the column name that will be used as the first description component.
#' @param dsc2 the column name that will be used as the second description component.
#' @param smr_col text color for the summary component. Defaults to 'white'.
#' @param smr_bgcol background-color for the summary component. Defaults to "#002240".
#' @param dsc_col text color for the first description component. Defaults to "#1D1D1D".
#' @param dsc_bgcol background-color for the first description component. Defaults to "#FFFFFF".
#' @param dsc_size font size of the first description component. Defaults to "14px".
#' @param dsc2_col text color for the second description component. Defaults to "#3AD900".
#' @param dsc2_bgcol background-color for the second description component. Defaults to "#FFFFFF".
#' @param dsc2_size font size of the second description component. Defaults to "14px".
#' @param open logical, whether to open by default the summary components. Default to FALSE

#' @return A Vertical HTML Time Line Widget with 2 Description Levels
#' @export
#'



create_tml_2 <- function(df,
                       smr,
                       dsc,
                       dsc2,
                       smr_col = "white",
                       smr_bgcol =  "#002240",
                       dsc_col = "#1D1D1D",
                       dsc_bgcol = "#FFFFFF",
                       dsc_size = "14px",
                       dsc2_col = "#1D1D1D",
                       dsc2_bgcol = "#FFFFFF",
                       dsc2_size = "14px",
                       open = FALSE) {

  if (any(!is.character(smr),
          !is.character(dsc),
          !is.character(dsc2),
          !is.character(smr_col),
          !is.character(smr_bgcol),
          !is.character(dsc_col),
          !is.character(dsc_bgcol),
          !is.character(dsc_size),
          !is.character(dsc2_col),
          !is.character(dsc2_bgcol),
          !is.character(dsc2_size))) {


    stop("All parameters except 'df' must be provided as a character string ! you're not in Tidyverse !!!")

  }


  if(any(!c(smr, dsc, dsc2) %in% names(df))) {

    stop(glue::glue(

      "Please check the columns' names.

      Note that they need to be provided as a character string in this form:

      (df = mtcars, smr = 'mpg', dsc = 'wt', dsc2 = 'hp')

      "

    ))

  }


  htmltools::tagList(

    cronologia_dependency(),

    timeline2(df,
              smr,
              dsc,
              dsc2,
              smr_col,
              smr_bgcol,
              dsc_col,
              dsc_bgcol,
              dsc_size,
              dsc2_col,
              dsc2_bgcol,
              dsc2_size,
              open)

  )

}


